mcp/src/rustic_ai/mcp/agent.py

```diff
@@ -0,0 +1,161 @@
+import asyncio
+from contextlib import AsyncExitStack
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+from mcp import ClientSession, StdioServerParameters
+from mcp.client.sse import sse_client
+from mcp.client.stdio import stdio_client
+from rustic_ai.core.guild import agent
+from rustic_ai.core.guild.agent import Agent, ProcessContext
+from rustic_ai.core.guild.agent_ext.mixins.guild_refresher import GuildRefreshMixin
+
+from .models import (
+    CallToolRequest,
+    CallToolResponse,
+    MCPAgentConfig,
+    MCPClientType,
+    MCPServerConfig,
+    ToolResult,
+)
+
+
+@dataclass
+class MCPSessionContext:
+    session: ClientSession
+    exit_stack: AsyncExitStack
+    client_type: str
+
+
+class MCPAgent(Agent[MCPAgentConfig], GuildRefreshMixin):
+    """
+    Agent that connects to MCP servers and exposes their capabilities.
+
+    This agent manages MCP client sessions. Due to asyncio event loop constraints,
+    sessions must be created and used within the same event loop context.
+    """
+
+    def __init__(self):
+        self._sessions: Dict[str, MCPSessionContext] = {}
+        self._connection_locks: Dict[str, asyncio.Lock] = {}
+
+    @property
+    def server_configs(self) -> Dict[str, MCPServerConfig]:
+        return {s.name: s for s in self.config.servers}
+
+    async def _initialize_session(
+        self, transport_context, config: MCPServerConfig, client_type: str
+    ) -> MCPSessionContext:
+        """Helper to initialize session from a transport context."""
+        exit_stack = AsyncExitStack()
+        try:
+            # Enter the transport context
+            read, write = await exit_stack.enter_async_context(transport_context)
+
+            # Enter the ClientSession context
+            session = await exit_stack.enter_async_context(ClientSession(read, write))
+
+            # Initialize the session
+            await session.initialize()
+
+            # Verify connection by listing tools
+            result = await session.list_tools()
+            tool_names = [tool.name for tool in result.tools]
+            self.logger.info(f"Connected to {config.name} ({client_type}). Available tools: {tool_names}")
+
+            return MCPSessionContext(session=session, exit_stack=exit_stack, client_type=client_type)
+
+        except Exception as e:
+            await exit_stack.aclose()
+            raise e
+
+    async def _connect_server(self, server_name: str) -> Optional[ClientSession]:
+        """Ensures connection to the specified server exists."""
+
+        # Fast path check
+        if server_name in self._sessions:
+            return self._sessions[server_name].session
+
+        config = self.server_configs.get(server_name)
+        if not config:
+            self.logger.error(f"Server configuration not found for: {server_name}")
+            return None
+
+        # Lock for this specific server to prevent race conditions
+        if server_name not in self._connection_locks:
```

Comment:
We process single message at a time, what will be the race condition? Also, we should have single server.